File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        if(root == null){
            return new ArrayList<>();
        }

        Deque<TreeNode> dq = new ArrayDeque<>();

        List<Integer> list = new ArrayList<>();

        dq.offerLast(root);

        while(!dq.isEmpty()){
            int size = dq.size();
            list.add(dq.peekLast().val);
            for(int i = 0; i < size; i++){
                TreeNode curr = dq.pollFirst();
                if(curr.left != null){
                    dq.offerLast(curr.left);
                }
                if(curr.right != null){
                    dq.offerLast(curr.right);
                }
            }
        }

        return list;
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(root == nullptr){
            return true;
        }

        int left = getHeight(root->left);
        int right = getHeight(root->right);

        if(abs(left - right) > 1){
            return false;
        }

        return isBalanced(root->left) && isBalanced(root->right);

    }
private:
    int getHeight(TreeNode* root){
        if(root == nullptr){
            return 0;
        }

        int left = getHeight(root->left) + 1;
        int right = getHeight(root->right) + 1;

        return max(left, right);
    }
};

    
